CORTX-34166: [cortx-ha] Consul transaction changes to address 64 KV limit - part #1

Madhura-08 opened the pull request:
…imit - part #1

consul transactions have limitations of pushing 64 keys/64 transactions in one request. Because of this, HA was failing on 27 node deployment setup because more than 64 keys creation request was getting executed in one transaction request.

Solution:
Divide the HA key creation into parts. First event manager subscription, cluster, site, and rack health keys which are fixed in numbers will be created in one request. For node health keys creation, every 64 keys creation request will be sent sequentially. If keys are less than that, it will be done in one request.

Signed-off-by: Madhura Mande <madhura.mande@seagate.com>

# Problem Statement
https://jts.seagate.com/browse/CORTX-34166)

# Design
Divide the HA key creation into parts. First event manager subscription, cluster, site, and rack health keys which are fixed in numbers will be created in one request. For node health keys creation, every 64 keys creation request will be sent sequentially. If keys are less than that, it will be done in one request.

# Coding
-  [ ] Coding conventions are followed and code is consistent

# Testing 
- [x] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [x] Testing was performed with RPM
https://jts.seagate.com/secure/attachment/535097/CORTX-34166_Test_details.txt

# Review Checklist 
- [x] PR is self reviewed
- [x] JIRA number/GitHub Issue added to PR
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained
- [ ] Is there a change in filename/package/module or signature? [Y/N]: 
- [ ] If yes for above point, is a notification sent to all other cortx components? [Y/N]
- [ ] Side effects on other features (deployment/upgrade)? [Y/N]
- [ ] Dependencies on other component(s)? [Y/N]
-     If yes for above point, post link to the corresponding PR.

# Review Checklist 
- [ ] Is perfline test run and the report with and without the changes updated in the PR? [Y/N]: 

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

Madhura-08 committed:
CORTX-34166: [cortx-ha] Consul transaction changes to address 64 KV limit - part #1

consul transactions has limitations of pushing 64 keys/64 transactions in one
request. Because of this, HA was failing on 27 node deployment setup because
more than 64 keys creation request was getting executed in one transaction
request.

Solution:
Divide the HA key creation in parts. First event manager subscription, cluster,
site, rack health keys which are fix in numbers will be created in one request.
For node health keys creation, every 64 keys creation request will be sent
sequentially. If keys less than that, it will be done in one request.

Signed-off-by: Madhura Mande <madhura.mande@seagate.com>

Madhura-08 committed:
CORTX-34166: [cortx-ha] Consul transaction changes to address 64 KV limit - part #1

- Add small comment fixes

Signed-off-by: Madhura Mande <madhura.mande@seagate.com>

d-nayak commented on ha/k8s_setup/ha_setup.py:
```diff
@@ -370,15 +372,26 @@ def _add_cluster_component_health(self) -> None:
+                    Log.debug(f"pushing {CONSUL_TRANSACTIONS_LIMIT} node health transactions to store")
+                    self._confstore.commit()
+                node_id_list.clear()
+                node_id_list.append(node_id)
```
Am I missing something here - why are we appending node_id 'again' to node_id_list?

Madhura-08 replied:
node_id is coming from the node_mapping dictionary. I have explicitly introduced the node_id list here for allowing 64 transactions in one consul transaction request.
If the length of this list exceeds (>64), This confstore.commit() will be called which will push 64 keys to kv store. (So, for every 64 bunch, this will happen).

d-nayak replied:
OK, make sense.

Madhura-08 replied:
Sure, Thanks!

seagate-sarang-sawant commented on ha/k8s_setup/ha_setup.py:
```diff
@@ -370,15 +372,26 @@ def _add_cluster_component_health(self) -> None:
+                # to push all the local cashed values to consul server
+                if kv_enable_batch_put:
+                    Log.debug(f"pushing {CONSUL_TRANSACTIONS_LIMIT} node health transactions to store")
+                    self._confstore.commit()
```
@Madhura-08 I am not sure if we are setting cardinality for subsequent batches after first batch. Unless we do that it seems that cardinality set earlier will be reused. If I am missing any internal way to populate payload please correct me.

Madhura-08 replied:
Cardinality will be stored in the first iteration itself. For cardinality, we have only one key. Then this stored cardinality will be reused here to get the machine-ids out of it. And then these machine-ids will be used to create node health keys.
Ex: 
cortx>ha>v2>cluster_cardinality:{"num_nodes": 10, "node_id_dict": {"cortx-data-g0-2.cortx-data-headless.cortx.svc.cluster.local": "6ac2b08cfe659acf910fed4d11b7df53"}}

So, In this key, **6ac2b08cfe659acf910fed4d11b7df53** is the machine-id. We will iterate this node_id_dict for its values and then the extracted machine id will be used to create the node health key.
Ex node health key:
**cortx>ha>v2>cortx>ha>system>cluster>b59c6099-0f46-4e31-9215-aad37bf838e1>site>NOT_DEFINED>rack>NOT_DEFINED>node>f488442b875fe8c4328e148f8aca7aa7>health**

I hope this will help.

seagate-sarang-sawant replied:
I understand, my question was more towards setting payload for batch commit. Setting cardinality seems to be the only way as per API.

mssawant commented:
commit message summary is too long, consider following points and can be fixed as follows,

1. No need to mention the repo.
2. Try to keep the summary short, problem is anyway described in details later.

```
CORTX-34166: Consul restricts KV updates per transaction

consul transactions has limitations of pushing 64 keys/64 transactions in one
request. Because of this, HA was failing on 27 node deployment setup because
more than 64 keys creation request was getting executed in one transaction
request.

Solution:
Divide the HA key creation in parts. First event manager subscription, cluster,
site, rack health keys which are fix in numbers will be created in one request.
For node health keys creation, every 64 keys creation request will be sent
sequentially. If keys less than that, it will be done in one request.

Signed-off-by: Madhura Mande <madhura.mande@seagate.com>
```

mssawant commented on ha/k8s_setup/ha_setup.py:
```diff
@@ -370,15 +372,26 @@ def _add_cluster_component_health(self) -> None:
         # list, then mark the respective functional type and update the specific info
         for node_id in node_mapping.values():
+            node_id_list.append(node_id)
+            if len(node_id_list) > CONSUL_TRANSACTIONS_LIMIT:
```
Instead of calculating the length of the list in every iteration, just maintain a separate counter and check that,
e.g.
```
nr_transactions = 0
for node_id in node_mapping.values():
    if nr_transactions < CONSUL_TRANSACTIONS_LIMIT:
        nr_transactions += 1
        continue
    nr_transactions = 0   
    if kv_enabl_batch_put:
        ...
```
or if you need node_id list, then you can use list comprehension directly, you already are calculating length node_mapping, so
e.g.
```
node_ids = node_mapping.values()
start = 0
while start < node_id_len:
    node_id_list_to_commit = node_ids[start: start + CONSUL_TRANSACTIONS_LIMIT]
    # commit
    start += CONSUL_TRANSACTIONS_LIMIT
```

Madhura-08 replied:
Instead of this, I am using enumeration now. Which will give an index also. And the same index can be used to track the number of transactions that are already part of batch put.

mssawant commented on ha/k8s_setup/ha_setup.py:
```diff
@@ -391,6 +404,12 @@ def _add_node_health(self, data_node_ids, server_node_ids, control_node_ids) ->
+            # to push all the local cashed values to consul server
+            if kv_enable_batch_put:
+                Log.debug(f"pushing node health transactions to store")
+                self._confstore.commit()
```
where is the batch populated?

Madhura-08 replied:
https://github.com/Seagate/cortx-ha/blob/f16a715a6f8079b93a125852d0c3fac4d918c305/ha/util/consul_kv_store.py#L330
This is commit and which will internally call batch put:
https://github.com/Seagate/cortx-ha/blob/f16a715a6f8079b93a125852d0c3fac4d918c305/ha/util/consul_kv_store.py#L287

mssawant commented on ha/k8s_setup/ha_setup.py:
```diff
@@ -391,6 +404,12 @@ def _add_node_health(self, data_node_ids, server_node_ids, control_node_ids) ->
                                    resource_type=CLUSTER_ELEMENTS.NODE.value,
                                    resource_id=node_id,
                                    specific_info=specific_info)
+        if node_id_list:
```
This can be re-written as `if node_id_list  and kv_enable_batch_put:`

Madhura-08 replied:
Yes. Thanks. instead of a list, I am using another flag now. So, checking that flag here.

Madhura-08 committed:
CORTX-34166: [cortx-ha] Consul transaction changes to address 64 KV limit

- Use dictionary enumeration instead of list for tracking number
  of keys that can be commited to consul transaction put

Signed-off-by: Madhura Mande <madhura.mande@seagate.com>

gauravchaudhari02 commented:
retest this please

shailesh-vaidya commented:
Closing as an obsolete
